### PR TITLE
Port Softmax kernel to OpenCL

### PR DESCRIPTION
**Pull request by xiaoqie**
Only float32 is working now on OpenCL.
I can't test on CUDA. Can anyone test this on CUDA?

### CODE REVIEW

**Comment by nouiz** [comment]
jenkins test this

this should test with CUDA.

File: theano/gpuarray/nnet.py [hunk 1]
```diff
@@ -607,71 +593,177 @@ def gpu_kernels(self, node, nodename):
...
+                }
+
+                local_barrier();
+                %(ctype)s row_sum = buf[0];
+                local_barrier();
+                for (ga_int tx = LID_0; tx< N; tx += LDIM_0){
+                    sm_ptr[tx * sm_s1] = %(write_sm)s(exp(%(load_x)s(x_ptr[tx * sx1]) - row_max) / row_sum);
+                }
+                local_barrier();
+                local_barrier();
```
Comment (abergeron) [inline theano/gpuarray/nnet.py]: You definitively don't need two in a row.

File: theano/gpuarray/nnet.py [hunk 2]
```diff
@@ -607,71 +593,177 @@ def gpu_kernels(self, node, nodename):
...
+                    }
+                    local_barrier();
+                    //reduce so that LID_0 0 has the reduction of everything
+                    for (ga_uint _n = GA_WARP_SIZE / 2; _n > 0; _n /= 2) {
+                        if (LID_0 < _n && LID_0 + _n < N)
+                            buf[LID_0] = max(buf[LID_0], buf[LID_0+_n]);
+                        local_barrier();
+                    }
+                }
+                local_barrier();
```
Comment (abergeron) [inline theano/gpuarray/nnet.py]: This is superfluous

File: theano/gpuarray/nnet.py [hunk 3]
```diff
@@ -607,71 +593,177 @@ def gpu_kernels(self, node, nodename):
...
+                    for (ga_uint _n = GA_WARP_SIZE / 2; _n > 0; _n /= 2) {
+                        if (LID_0 < _n && LID_0 + _n < N)
+                            buf[LID_0] = buf[LID_0] + buf[LID_0+_n];
+                        local_barrier();
+                    }
+                }
+
+                local_barrier();
+                %(ctype)s row_sum = buf[0];
+                local_barrier();
```
Comment (abergeron) [inline theano/gpuarray/nnet.py]: Both local barriers around here are redundant

File: theano/gpuarray/nnet.py [hunk 4]
```diff
@@ -821,76 +911,184 @@ def gpu_kernels(self, node, nodename):
...
+                    local_barrier();
+                    //reduce so that LID_0 0 has the reduction of everything
+                    for (ga_uint _n = GA_WARP_SIZE / 2; _n > 0; _n /= 2) {
+                        if (LID_0 < _n && LID_0 + _n < N)
+                            buf[LID_0] = max(buf[LID_0], buf[LID_0+_n]);
+                        local_barrier();
+                    }
+                }
+
+                local_barrier();
```
Comment (abergeron) [inline theano/gpuarray/nnet.py]: This is not required

File: theano/gpuarray/nnet.py [hunk 5]
```diff
@@ -821,76 +911,184 @@ def gpu_kernels(self, node, nodename):
...
+                    local_barrier();
+                    //reduce so that LID_0 0 has the reduction of everything
+                    for (ga_uint _n = GA_WARP_SIZE / 2; _n > 0; _n /= 2) {
+                        if (LID_0 < _n && LID_0 + _n < N)
+                            buf[LID_0] = buf[LID_0] + buf[LID_0+_n];
+                        local_barrier();
+                    }
+                }
+
+                local_barrier();
```
Comment (abergeron) [inline theano/gpuarray/nnet.py]: This is not required

File: theano/gpuarray/nnet.py [hunk 6]
```diff
@@ -821,76 +911,184 @@ def gpu_kernels(self, node, nodename):
...
+                    local_barrier();
+                    //reduce so that LID_0 0 has the reduction of everything
+                    for (ga_uint _n = GA_WARP_SIZE / 2; _n > 0; _n /= 2) {
+                        if (LID_0 < _n && LID_0 + _n < N)
+                            buf[LID_0] = max(buf[LID_0], buf[LID_0+_n]);
+                        local_barrier();
+                    }
+                }
+
+                local_barrier();
```
Comment (abergeron) [inline theano/gpuarray/nnet.py]: Superfluous

File: theano/gpuarray/nnet.py [hunk 7]
```diff
@@ -821,76 +911,184 @@ def gpu_kernels(self, node, nodename):
...
+                    local_barrier();
+                    //reduce so that LID_0 0 has the reduction of everything
+                    for (ga_uint _n = GA_WARP_SIZE / 2; _n > 0; _n /= 2) {
+                        if (LID_0 < _n && LID_0 + _n < N)
+                            buf[LID_0] = buf[LID_0] + buf[LID_0+_n];
+                        local_barrier();
+                    }
+                }
+
+                local_barrier();
```
Comment (abergeron) [inline theano/gpuarray/nnet.py]: Superfluous

File: theano/gpuarray/nnet.py [hunk 8]
```diff
@@ -821,76 +911,184 @@ def gpu_kernels(self, node, nodename):
...
+                %(ctype)s row_sum = buf[0];
+                local_barrier();
+                for (ga_int tx = LID_0; tx< N; tx += LDIM_0){
+                    sm_ptr[tx * sm_s1] = %(write_sm)s(exp(%(load_x)s(x_ptr[tx * sx1]) + %(load_b)s(b[tx * sb0]) - row_max) / row_sum);
+                }
+                local_barrier();
+                local_barrier();
```
Comment (abergeron) [inline theano/gpuarray/nnet.py]: No need for two

**Comment by abergeron** [comment]
jenkins test this please

**Comment by xiaoqie** [comment]
jenkins test this please

**Comment by nouiz** [comment]
jenkins test this

@xiaoqie you don't have the authorization to trigger jenkins.

**Comment by xiaoqie** [comment]
@nouiz @abergeron Can you trigger jenkins please?

**Comment by abergeron** [comment]
jenkins test this please

**Pull request merged**